src/visual.ts

```diff
@@ -260,122 +273,163 @@ export class Visual implements IVisual {
+        const hasHighlights: boolean = !!valuesColumn.highlights;
 
-        let minSize: number = settings.dots.minSize;
-        let maxSize: number = settings.dots.maxSize;
-        if (settings.dots) {
-            minSize = settings.dots.minSize;
-            maxSize = settings.dots.maxSize;
-        }
+        const options: SeriesGenerationOptions = { categoryValues, grouped, settings, columns, timeStampColumn, host, maxCategoryValue, minCategoryValue, dataPointLabelSettings, isScalar, runnerCounterFormatString, hasHighlights, valuesColumn, behavior };
+        const series: Series[] = Visual.generateSeries(options);
 
-        let eventSizeScale: LinearScale = <LinearScale>Visual.createScale(
-            true,
-            columns.EventSize ? [d3Min(<number[]>columns.EventSize.values), d3Max(<number[]>columns.EventSize.values)] : [0, 0],
-            minSize,
-            maxSize);
+        const xAxisCardProperties = Helpers.getCategoryAxisProperties(dataView.metadata);
+        const valueAxisProperties = Helpers.getValueAxisProperties(dataView.metadata);
+        const valuesMetadataArray: powerbiVisualsApi.DataViewMetadataColumn[] = Visual.extractValuesMetadata(dataView);
 
-        let xAxisCardProperties: DataViewObject = Helpers.getCategoryAxisProperties(dataView.metadata);
+        const axesLabels = Visual.createAxesLabels(xAxisCardProperties, valueAxisProperties, timeStampColumn.source, valuesMetadataArray);
+        const dots: DataPoint[] = Visual.getDataPointsFromSeries(series);
 
-        let hasDynamicSeries: boolean = !!(timeStampColumn.values && timeStampColumn.source);
-        let hasHighlights: boolean = !!valuesColumn.highlights;
+        behavior.applySelectionStateToData(dots);
 
-        let dataPointLabelSettings = PulseChartDataLabelUtils.getDefaultPulseChartLabelSettings();
-        let gapWidths = Visual.getGapWidths(categoryValues);
-        let maxGapWidth = Math.max.apply(null, gapWidths);
+        return {
+            columns: columns,
+            dots: dots,
+            series: series,
+            isScalar: isScalar,
+            dataLabelsSettings: dataPointLabelSettings,
+            axesLabels: { x: axesLabels.xAxisLabel, y: axesLabels.yAxisLabel },
+            hasDynamicSeries: hasDynamicSeries,
+            categoryMetadata: timeStampColumn.source,
+            categories: categoryValues,
+            settings: settings,
+            grouped: grouped,
+            hasHighlights: !!valuesColumn.highlights,
+            widthOfXAxisLabel: Visual.DefaultXAxisLabelWidth,
+            widthOfTooltipValueLabel: widthOfTooltipValueLabel,
+            heightOfTooltipDescriptionTextLine: heightOfTooltipDescriptionTextLine,
+            runnerCounterHeight: textMeasurementService.measureSvgTextHeight(
+                Visual.GET_RUNNER_COUNTER_TEXT_PROPERTIES("lj", settings.runnerCounter.fontSize.value))
```

Comment:
We have multiple functions doing the same thing - generate TextProperties. Please check methods from line 171-202. Replace them with one more universal function. 
_Example:_
```
public static getTextProperties(
        { text = "", fontFamily = "sans-serif", fontSizeValue = 12, fontWeight = "normal" }: 
        { text?: string, fontFamily?: string, fontSizeValue?: number, fontWeight?: string } = {}
): TextProperties {
        return {
            text: text,
            fontFamily: fontFamily,
            fontSize: fontSizeValue + "px",
            fontWeight: fontWeight
  };
}
```

Reply:
Done.